subs/pantry/src/Pantry/Hackage.hs

```diff
@@ -463,11 +487,22 @@ getHackageTarball
   => PackageIdentifierRevision
   -> Maybe TreeKey
   -> RIO env Package
-getHackageTarball pir@(PackageIdentifierRevision name ver _cfi) mtreeKey = do
+getHackageTarball = getHackageTarballOnGPD (\ _ _ -> pure ())
+
+-- | Same as `getHackageTarball`, but allows an extra action to be performed on the parsed
+-- `GenericPackageDescription` and newly created `TreeId`.
+getHackageTarballOnGPD
+  :: (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+  => (TreeId -> GenericPackageDescription -> RIO env ())
```

Comment:
Alternative approach: what about having a `HackageTarballResult` datatype that contains a `Package`, `TreeId`, and `GenericPackageDescription` and return that? Any specific reason why this callback should be called inside this function call?

Reply:
Correct me if I am wrong, but from my understanding `getHackageTarball` does not need to retrieve `GenericPackageDescription`, it only needs the `Package`.  `withCachedTree` allows for the direct  look up of the `Package` from the database if such is present, and only falling back onto building the `Package` by pulling it from Hackage. In other words that function `onGPD`, will only be called when this `Package` is new to Pantry.

To answer your question. We could add new datatype `HackageTarballResult` that would get all three: `Package`, `TreeId`, and `GenericPackageDescription`, but they will be used only in `stackage-server`, for Pantry pulling the cabal blob from database and parsing it will be an unnecessary overhead.

Reply:
If `getHackageTarball` doesn't need the `GenericPackageDescription`, and `getHackageTarball` is now implemented in terms of `getHackageTarballOnGPD`, that would mean that we are unnecessarily parsing the cabal file.

> To answer your question. We could add new datatype HackageTarballResult that would get all three: Package, TreeId, and GenericPackageDescription, but they will be used only in stackage-server, for Pantry pulling the cabal blob from database and parsing it will be an unnecessary overhead.

I think I'd prefer that over

1. Adding a new function to the API and
2. Using callback style where it isn't necessary

Reply:
I did add a new `HackageTarballResult` data type, which it is not part of a public api and is exported through an internal module to stackage server. I think this is a win win solution, since it avoids all three: unnecessary overhead of parsing cabal file when it is not needed, the callback style and addition of another function to the api.